### PR TITLE
Implemented Howard`s algorithm for finding minimum cycle mean

### PR DESCRIPTION
**Pull request by SChudakov**
Implemented Howard\`s algorithm for finding minimum cycle mean as discussed in #861.

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HowardMinimumMeanCycle.java
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HowardMinimumMeanCycle.java]: Not to start a full review, but could you add a few lines describing what a 'minimum cycle mean' is in a graph?

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MinimumCycleMeanAlgorithm.java
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MinimumCycleMeanAlgorithm.java]: Are there alternative algorithms that compute a minimumCycleMean? If not, do we really need an interface? (I had never heard about a Minimum Cycle Mean before seeing this PR, so please excuse my ignorance).

Comment (SChudakov) [reply jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MinimumCycleMeanAlgorithm.java]: The idea to implement this algorithm was derived from the project`s wiki page. There are two other algorithms mentioned there, in particular the Karp`s algorithm and the Hartman-Orlin`s algorithm. Furthermore, in the article several more algorithms are mentioned.

I decided to start with this one because it is reported to be the fastest one. Never the less its complexity is exponential. The Karp`s and the Hartman-Orlin`s algorithms have polynomial complexities in terms of number of vertices and number of edges. So there is a theoretical improvement there.

Does it make sense to you if we had additionally an implementation of a polynomial algorithm for this problem even though it possibly not going to be faster in most of the cases? If so, I could implement it in this PR with a corresponding performance comparison.

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HowardMinimumMeanCycle.java
Comment (SChudakov) [reply jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HowardMinimumMeanCycle.java]: There is a formal definition of this property in the interface for the algorithm.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MinimumCycleMeanAlgorithm.java
Comment (Toptachamann) [reply jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MinimumCycleMeanAlgorithm.java]: Potentially we can implement a polynomial algorithm for this problem, so this interface makes sense to me. However, I would prefer not to add it in this PR as it will complicated the review process

Comment (SChudakov) [reply jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MinimumCycleMeanAlgorithm.java]: ok

**Comment by Toptachamann** [comment]
General comment from me would be to add some inline code comments to make the implementation more readable

**Comment by d-michail** [comment]
@jkinable I remember that there is an algorithm which computes minimum cost flows by starting with any feasible flow and repeatedly sending flow along such a cycle (mean cycle cancelling algorithm).

**Comment by Toptachamann** [comment]
@d-michail I think it’s called Goldberg-Tarjan algorithm described in {Network flows. Ahuja et al. chapter 10.5}

**Comment by SChudakov** [comment]
I\`ve addressed issues made by @Toptachamann  and @d-michail in the latest commit.

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HowardMinimumMeanCycle.java
Comment (SChudakov) [reply jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HowardMinimumMeanCycle.java]: In fact, the previous version of the algorithm used the `queue` variable twice in the way which makes the suggested change impossible. In the latest commit I\`ve changed the implementation to make it closer to the pseudocode in the original article. This made it possible to use  `Dequeue` interface and `ArrayDeque` implementation correspondingly.

**Comment by d-michail** [comment]
@Toptachamann @jkinable Any more comments on this one? If not I will merge in a couple of days.

**Review by Toptachamann — CHANGES_REQUESTED** [review]
Looks good to me. The main point is to improve tests by removing unnecessary data + reuse existing util methods. After that, this PR can finally get merged.

**Comment by SChudakov** [comment]
Done

**Review by Toptachamann — APPROVED** [review]
Looks good to me, great job! This PR can be merged soon unless there're more comments.

**Comment by d-michail** [comment]
Thanks! Great job!

**Pull request merged**
